File: SDT/Code/sdt_navier_cpp/src/solver.cpp
#include "sdt_navier/solver.hpp"
#include "sdt_navier/constants.hpp"
#include "sdt_navier/operators.hpp"
#include "sdt_navier/fields.hpp"
#include <algorithm>
#include <cmath>
#include <functional>

namespace sdt_navier {

SDTNavierSolver::SDTNavierSolver(
    FieldSystem& fields,
    const SDTNavierEquations& equations,
    double dt,
    double cfl,
    const std::string& method,
    bool enforce_incompressibility
) : fields_(fields), equations_(equations), cfl_(cfl),
    method_(method), enforce_incompressibility_(enforce_incompressibility),
    t_(fields.t())
{
    if (dt == 0.0) {
        dt_ = estimate_timestep();
    } else {
        dt_ = dt;
    }
}

double SDTNavierSolver::estimate_timestep() const {
    // CFL condition: dt < CFL * dx / |v_max|
    double v_max = 0.0;
    for (const auto& vi : fields_.v()) {
        double v_mag = vi.norm();
        if (v_mag > v_max) v_max = v_mag;
    }
    if (v_max == 0.0) v_max = 1.0;

    double dx_min = std::min({fields_.dx(), fields_.dy(), fields_.dz()});
    double dt = cfl_ * dx_min / v_max;

    // Also consider slip damping timescale
    auto eta = fields_.eta();
    double eta_max = *std::max_element(eta.begin(), eta.end());
    if (eta_max > 0.0) {
        double dt_slip = 1.0 / (equations_.beta_slip() * eta_max / equations_.rho_s());
        dt = std::min(dt, dt_slip);
    }

    return dt;
}

void SDTNavierSolver::step() {
    if (method_ == "euler") {
        step_euler();
    } else if (method_ == "rk4") {
        step_rk4();
    }
    fields_.set_t(t_);
    t_ += dt_;
}

void SDTNavierSolver::step_euler() {
    // Compute gradients
    auto P_vec = fields_.P();
    std::vector<double> P(P_vec.data(), P_vec.data() + fields_.size());
    auto grad_P = compute_gradient(P, fields_, "extrapolate");

    auto kappa_vec = fields_.kappa();
    std::vector<double> kappa(kappa_vec.data(), kappa_vec.data() + fields_.size());
    auto grad_kappa = compute_gradient(kappa, fields_, "extrapolate");

    auto eta_vec = fields_.eta();
    std::vector<double> eta(eta_vec.data(), eta_vec.data() + fields_.size());
    auto grad_eta = compute_gradient(eta, fields_, "extrapolate");

    auto e_vec = fields_.e();
    std::vector<double> e(e_vec.data(), e_vec.data() + fields_.size());
    auto grad_e = compute_gradient(e, fields_, "extrapolate");

    auto grad_v = compute_velocity_gradient(fields_.v(), fields_, "extrapolate");

    // Compute force functionals
    auto F_curv = equations_.compute_force_curvature(fields_, grad_kappa);
    auto F_slip = equations_.compute_force_slip(fields_);
    auto C = equations_.compute_curvature_creation(fields_, grad_v);
    auto D = equations_.compute_curvature_destruction(fields_);
    auto S_strain = equations_.compute_slip_strain(fields_, grad_v);
    auto S_healing = equations_.compute_slip_healing(fields_);

    // Compute RHS
    auto dv_dt = equations_.compute_flow_rhs(fields_, grad_P, grad_v, F_curv, F_slip);
    auto dkappa_dt = equations_.compute_curvature_rhs(fields_, grad_kappa, grad_v, C, D);
    auto deta_dt = equations_.compute_slip_rhs(fields_, grad_eta, grad_v, S_strain, S_healing);

    auto sigma = compute_diversion_density(fields_);
    auto de_dt = equations_.compute_energy_rhs(fields_, grad_e, sigma);

    // Update fields
    auto& v = fields_.v();
    for (std::size_t i = 0; i < fields_.size(); ++i) {
        v[i] += dt_ * dv_dt[i];
    }

    auto kappa_map = fields_.kappa();
    for (std::size_t i = 0; i < fields_.size(); ++i) {
        kappa_map[i] += dt_ * dkappa_dt[i];
    }

    auto eta_map = fields_.eta();
    for (std::size_t i = 0; i < fields_.size(); ++i) {
        eta_map[i] += dt_ * deta_dt[i];
        eta_map[i] = std::clamp(eta_map[i], 0.0, 1.0);
    }

    auto e_map = fields_.e();
    for (std::size_t i = 0; i < fields_.size(); ++i) {
        e_map[i] += dt_ * de_dt[i];
    }

    if (enforce_incompressibility_) {
        project_pressure();
    }
}

void SDTNavierSolver::step_rk4() {
    // Simplified RK4: store intermediate states and average
    // Full implementation would recompute RHS at each stage
    step_euler();  // For now, use Euler (can be enhanced)
}

void SDTNavierSolver::project_pressure() {
    // Simplified pressure projection
    auto div_v = compute_divergence(fields_.v(), fields_, "extrapolate");

    // Adjust pressure
    auto P = fields_.P();
    double alpha_p = 1.0e10;
    for (std::size_t i = 0; i < fields_.size(); ++i) {
        P[i] += alpha_p * div_v[i] * dt_;
    }

    // Correct velocity
    auto grad_div = compute_gradient(div_v, fields_, "extrapolate");
    auto& v = fields_.v();
    double beta_v = 0.1 * dt_;
    for (std::size_t i = 0; i < fields_.size(); ++i) {
        v[i] -= beta_v * FieldSystem::Vector3d(grad_div[i][0], grad_div[i][1], grad_div[i][2]);
    }
}

void SDTNavierSolver::run_until(
    double t_end,
    std::function<void(SDTNavierSolver&)> callback
) {
    while (t_ < t_end) {
        if (t_ + dt_ > t_end) {
            dt_ = t_end - t_;
        }
        step();
        if (callback) {
            callback(*this);
        }
    }
}

double SDTNavierSolver::get_divergence_error() const {
    auto div_v = compute_divergence(fields_.v(), fields_, "extrapolate");
    double max_div = 0.0;
    for (double div : div_v) {
        max_div = std::max(max_div, std::abs(div));
    }
    return max_div;
}

}  // namespace sdt_navier

File: SDT/Code/sdt_navier/solver.py
"""
SDT-Navier Solver Kernel

Implements time-stepping for SDT-Navier field equations with:
- Explicit time-stepping (Euler or Runge-Kutta)
- Incompressibility enforcement via pressure projection
- Adaptive timestep based on CFL condition
"""

import numpy as np
from typing import Optional, Callable

try:  # SciPy is optional for the simplified projection used here
    from scipy.sparse import csc_matrix  # type: ignore
    from scipy.sparse.linalg import spsolve  # type: ignore
except ImportError:  # pragma: no cover - fallback for lightweight environments
    csc_matrix = None  # type: ignore
    spsolve = None  # type: ignore

from .fields import FieldSystem
from .equations import (
    SDTNavierEquations,
    compute_force_curvature,
    compute_force_slip,
    compute_curvature_creation,
    compute_curvature_destruction,
    compute_slip_strain,
    compute_slip_healing,
    compute_diversion_density,
)
from .lattice import (
    compute_gradient,
    compute_divergence,
    compute_velocity_gradient,
)


class SDTNavierSolver:
    """
    Solver for SDT-Navier field equations.
    """
    
    def __init__(
        self,
        fields: FieldSystem,
        equations: SDTNavierEquations,
        dt: Optional[float] = None,
        cfl: float = 0.5,
        method: str = "euler",
        enforce_incompressibility: bool = True,
    ):
        """
        Initialize SDT-Navier solver.
        
        Parameters
        ----------
        fields : FieldSystem
            Initial field state
        equations : SDTNavierEquations
            SDT-Navier equations with force functional parameters
        dt : float, optional
            Fixed timestep (s). If None, use adaptive timestep based on CFL.
        cfl : float
            CFL number for adaptive timestep (default 0.5)
        method : str
            Time-stepping method: "euler" or "rk2"
        enforce_incompressibility : bool
            Whether to enforce ∇·v = 0 via pressure projection
        """
        self.fields = fields
        self.equations = equations
        self.cfl = cfl
        self.method = method
        self.enforce_incompressibility = enforce_incompressibility
        
        if dt is None:
            # Estimate initial timestep from CFL condition
            dt = self._estimate_timestep()
        
        self.dt = dt
        self.t = fields.t
    
    def _estimate_timestep(self) -> float:
        """
        Estimate timestep from CFL condition: dt < CFL * dx / |v_max|.
        
        Returns
        -------
        dt : float
            Estimated timestep (s)
        """
        v_max = np.max(np.abs(self.fields.v))
        if v_max == 0:
            v_max = 1.0  # Default to avoid division by zero
        
        dx_min = min(self.fields.dx, self.fields.dy, self.fields.dz)
        dt = self.cfl * dx_min / v_max
        
        # Also consider diffusion timescale
        # For slip damping: dt < 1 / (beta_slip * eta_max / rho_s)
        eta_max = np.max(self.fields.eta)
        if eta_max > 0:
            dt_slip = 1.0 / (self.equations.beta_slip * eta_max / self.equations.rho_s)
            dt = min(dt, dt_slip)
        
        return dt
    
    def step(self) -> None:
        """
        Perform one time step.
        """
        if self.method == "euler":
            self._step_euler()
        elif self.method == "rk2":
            self._step_rk2()
        else:
            raise ValueError(f"Unknown method: {self.method}")
        
        # Update time
        self.fields.t = self.t
        self.t += self.dt
    
    def _step_euler(self) -> None:
        """
        Euler time-stepping: u^{n+1} = u^n + dt * du/dt.
        """
        # Compute gradients
        grad_P = compute_gradient(self.fields.P, self.fields, boundary="extrapolate")
        grad_kappa = compute_gradient(self.fields.kappa, self.fields, boundary="extrapolate")
        grad_eta = compute_gradient(self.fields.eta, self.fields, boundary="extrapolate")
        grad_e = compute_gradient(self.fields.e, self.fields, boundary="extrapolate")
        grad_v = compute_velocity_gradient(self.fields.v, self.fields, boundary="extrapolate")
        
        # Compute force functionals
        F_curv = compute_force_curvature(
            self.fields,
            grad_kappa,
            alpha_curv=self.equations.alpha_curv,
        )
        F_slip = compute_force_slip(
            self.fields,
            beta_slip=self.equations.beta_slip,
        )
        
        C = compute_curvature_creation(
            self.fields,
            grad_v,
            gamma_create=self.equations.gamma_create,
        )
        D = compute_curvature_destruction(
            self.fields,
            delta_destroy=self.equations.delta_destroy,
        )
        
        S_strain = compute_slip_strain(
            self.fields,
            grad_v,
            epsilon_strain=self.equations.epsilon_strain,
        )
        S_healing = compute_slip_healing(
            self.fields,
            zeta_heal=self.equations.zeta_heal,
        )
        
        # Compute RHS for each equation
        dv_dt = self.equations.compute_flow_rhs(
            self.fields,
            grad_P,
            grad_v,
            F_curv,
            F_slip,
        )
        
        dkappa_dt = self.equations.compute_curvature_rhs(
            self.fields,
            grad_kappa,
            grad_v,
            C,
            D,
        )
        
        deta_dt = self.equations.compute_slip_rhs(
            self.fields,
            grad_eta,
            grad_v,
            S_strain,
            S_healing,
        )
        
        sigma = compute_diversion_density(self.fields)
        de_dt = self.equations.compute_energy_rhs(
            self.fields,
            grad_e,
            sigma,
        )
        
        # Update fields
        self.fields.v += self.dt * dv_dt
        self.fields.kappa += self.dt * dkappa_dt
        self.fields.eta += self.dt * deta_dt
        self.fields.e += self.dt * de_dt
        
        # Clamp slip field to [0, 1]
        self.fields.eta = np.clip(self.fields.eta, 0.0, 1.0)
        
        # Enforce incompressibility
        if self.enforce_incompressibility:
            self._project_pressure()
    
    def _step_rk2(self) -> None:
        """
        Second-order Runge-Kutta (Heun's method).
        """
        # Store initial state
        v0 = self.fields.v.copy()
        kappa0 = self.fields.kappa.copy()
        eta0 = self.fields.eta.copy()
        e0 = self.fields.e.copy()
        
        # First stage (Euler step)
        self._step_euler()
        
        # Store intermediate state
        v1 = self.fields.v.copy()
        kappa1 = self.fields.kappa.copy()
        eta1 = self.fields.eta.copy()
        e1 = self.fields.e.copy()
        
        # Restore initial state
        self.fields.v = v0
        self.fields.kappa = kappa0
        self.fields.eta = eta0
        self.fields.e = e0
        
        # Second stage: compute RHS at intermediate state
        # (This is a simplified RK2 - full implementation would recompute RHS)
        # For now, use average of initial and intermediate
        self.fields.v = 0.5 * (v0 + v1)
        self.fields.kappa = 0.5 * (kappa0 + kappa1)
        self.fields.eta = 0.5 * (eta0 + eta1)
        self.fields.e = 0.5 * (e0 + e1)
        
        # Enforce incompressibility
        if self.enforce_incompressibility:
            self._project_pressure()
    
    def _project_pressure(self) -> None:
        """
        Enforce incompressibility ∇·v = 0 via proper pressure projection.
        
        Standard projection method:
        1. Compute divergence of intermediate velocity: div_v = ∇·v*
        2. Solve Poisson equation: ∇²φ = (1/Δt) ∇·v*
        3. Correct velocity: v^{n+1} = v* - ∇φ
        
        This ensures ∇·v^{n+1} = 0 exactly (up to numerical error).
        """
        # Step 1: Compute divergence of current velocity
        div_v = compute_divergence(self.fields.v, self.fields, boundary="extrapolate")
        
        # Step 2: Solve Poisson equation ∇²φ = (1/Δt) div_v
        # For periodic BCs, use FFT; for Dirichlet, use sparse solver
        # For now, implement a simple iterative solver (Jacobi) for Dirichlet BCs
        
        # Initialize potential
        phi = np.zeros_like(div_v)
        
        # Jacobi iteration for Poisson: ∇²φ = f
        # φ^{n+1} = (1/6) * (φ_neighbors + h² * f)
        # where h = dx (assuming cubic grid)
        h = self.fields.dx  # Grid spacing
        f = div_v / self.dt  # Source term: (1/Δt) div_v
        
        # Iterate until convergence (or fixed number of iterations)
        max_iter = 50
        tolerance = 1e-6
        
        for iteration in range(max_iter):
            phi_new = phi.copy()
            
            # Interior points: Jacobi update
            if self.fields.nx > 2 and self.fields.ny > 2 and self.fields.nz > 2:
                # 6-point stencil for 3D Laplacian
                phi_new[1:-1, 1:-1, 1:-1] = (
                    phi[2:, 1:-1, 1:-1] + phi[:-2, 1:-1, 1:-1] +  # x neighbors
                    phi[1:-1, 2:, 1:-1] + phi[1:-1, :-2, 1:-1] +  # y neighbors
                    phi[1:-1, 1:-1, 2:] + phi[1:-1, 1:-1, :-2] +  # z neighbors
                    h**2 * f[1:-1, 1:-1, 1:-1]
                ) / 6.0
            
            # Check convergence
            error = np.max(np.abs(phi_new - phi))
            phi = phi_new
            
            if error < tolerance:
                break
        
        # Step 3: Correct velocity: v^{n+1} = v* - ∇φ
        grad_phi = compute_gradient(phi, self.fields, boundary="extrapolate")
        self.fields.v -= grad_phi
        
        # Optional: Also update pressure field
        # P^{n+1} = P^n + φ (pressure increment)
        self.fields.P += phi
    
    def run_until(self, t_end: float, callback: Optional[Callable] = None) -> None:
        """
        Run simulation until t_end.
        
        Parameters
        ----------
        t_end : float
            End time (s)
        callback : callable, optional
            Function to call after each step: callback(solver)
        """
        while self.t < t_end:
            # Adjust timestep if needed
            if self.t + self.dt > t_end:
                self.dt = t_end - self.t
            
            self.step()
            
            if callback is not None:
                callback(self)
    
    def get_divergence_error(self) -> float:
        """
        Compute maximum divergence error: max|∇·v|.
        
        Returns
        -------
        error : float
            Maximum absolute divergence
        """
        div_v = compute_divergence(self.fields.v, self.fields, boundary="extrapolate")
        return np.max(np.abs(div_v))

